Update doc (#7740)
dotnet/msbuild: documentation/MSBuild-Server.md

# MSBuild Server

> MSBuild Server is basically an another type of node which can accept build request from clients and utilize worker nodes in current fashion to build projects. Main purpose of the server node is to avoid expensive MSBuild process start during build from tools like .NET SDK,
> and preserve caches between builds.

Reviewer: ```suggestion
MSBuild Server nodes accept build requests from clients and use worker nodes in the current fashion to build projects. The main purpose of the server node is to preserve caches between builds and avoid expensive MSBuild process start operations during build from tools like the .NET SDK.
```

## Usage

> The primary ways to use MSBuild are via Visual Studio and via CLI using `dotnet build`/`dotnet msbuild` commands. MSBuild Server is not supported in Visual Studio, becuase Visual studio itself
> works like MSBuild Server. For CLI is the server functionality enabled by default and can be disabled by setting `DOTNET_CLI_DO_NOT_USE_MSBUILD_SERVER` environment variable to value `1`.

Reviewer: ```suggestion
The primary ways to use MSBuild are via Visual Studio and via the CLI using the `dotnet build`/`dotnet msbuild` commands. MSBuild Server is not supported in Visual Studio because Visual studio itself works like MSBuild Server. For the CLI, the server functionality is enabled by default and can be disabled by setting the `DOTNET_CLI_DO_NOT_USE_MSBUILD_SERVER` environment variable to value `1`.
```

> In order to re-enable using of MSBuild Server, remove the variable or set the value to `0`.

Reviewer: ```suggestion
To re-enable MSBuild Server, remove the variable or set its value to `0`.
```

## Communication protocol

The server node uses same IPC approach as current worker nodes - named pipes. This solution allows to reuse existing code. When process starts, pipe with deterministic name is opened and waiting for commands. Client has following worfklow:

1. Try to connect to server
   - If server is not running, start new instance

>    - If server is busy or the connection is broken, fallback to classic build 

Reviewer: ```suggestion
   - If server is busy or the connection is broken, fall back to previous build behavior
```

2. Initiate handshake
2. Issue build command with `ServerNodeBuildCommand` packet
3. Read packets from pipe

>    - When `ServerNodeConsoleWrite` packet type is recieved, write content to appropriate output stream with respected coloring
>    - When `ServerNodeBuildResult` packet type is recieved, build is done and client writes trace message with exit code

Reviewer: ```suggestion
   - Write content to the appropriate output stream (respecting coloring) with the `ServerNodeConsoleWrite` packet
   - After the build completes, the `ServerNodeBuildResult` packet indicates the exit code [and provides other information? Not sure what else]
```
?

| BuildProcessEnvironment  | IDictionary<String, String>  | Environment variables for current build |
| Culture                  | CultureInfo                  | The culture value for current build |
| UICulture                | CultureInfo                  | The UI culture value for current build |
| ConsoleConfiguration     | TargetConsoleConfiguration   | Console configuration of target Console at which the output will be rendered |

`ServerNodeConsoleWrite` contains information for console output.

| Property name            | Type          | Description |
|---|---|---|
| Text                     | String        | The text that is written to the output stream. It includes ANSI escape codes for formatting. |
| OutputType               | Byte          | Identification of the output stream (1 = standard output, 2 = error output) |

> `ServerNodeBuildResult` informs about finished build.

Reviewer: ```suggestion
`ServerNodeBuildResult` indicates how the build finished.
```
